File: src/main/java/frc/robot/Robot.java
package frc.robot;
import edu.wpi.first.wpilibj.Timer;
import com.ctre.phoenix.motorcontrol.*;
import com.ctre.phoenix.motorcontrol.can.*;
import edu.wpi.first.wpilibj.Joystick;
import edu.wpi.first.wpilibj.TimedRobot;
import edu.wpi.first.networktables.NetworkTable;
import edu.wpi.first.networktables.NetworkTableEntry;
import edu.wpi.first.networktables.NetworkTableInstance;
import edu.wpi.first.cameraserver.CameraServer;

import edu.wpi.first.wpilibj.smartdashboard.SendableChooser;
import edu.wpi.first.wpilibj.smartdashboard.SmartDashboard;
import edu.wpi.first.wpilibj.motorcontrol.MotorController;
// import edu.wpi.first.wpilibj.SpeedControllerGroup;
import edu.wpi.first.wpilibj.motorcontrol.MotorControllerGroup;
import edu.wpi.first.wpilibj.XboxController;
import edu.wpi.first.wpilibj.GenericHID;
import edu.wpi.first.wpilibj.drive.DifferentialDrive;
import edu.wpi.first.networktables.*;


import edu.wpi.first.wpilibj.smartdashboard.SmartDashboard;

public class Robot extends TimedRobot {
    // Motor controllers

    private static final String kDefaultAuto = "Default";
    private static final String kCustomAuto = "My Auto";

    private String m_autoSelected;
    private final SendableChooser<String> m_chooser = new SendableChooser<>();


    TalonSRX _talon0 = new TalonSRX(1);
    TalonSRX _talon1 = new TalonSRX(3);

    TalonSRX _talon2 = new TalonSRX(5);
    TalonSRX _talon3 = new TalonSRX(2);
    TalonSRX _talon4 = new TalonSRX(4);
    TalonSRX _talon5 = new TalonSRX(6);

    private MotorController _leftMotors = new MotorController() {
        @Override
        public void set(double speed) {
            _talon0.set(ControlMode.PercentOutput, speed);
            _talon1.set(ControlMode.PercentOutput, speed);
            _talon1.set(ControlMode.PercentOutput, speed);
        }

        @Override
        public void setInverted(boolean isInverted) {
            _talon0.setInverted(false);
            _talon1.setInverted(false);
            _talon2.setInverted(false);
        }

        @Override
        public boolean getInverted() {
            return _talon0.getInverted();
        }

        @Override
        public void disable() {
        }

        @Override
        public void stopMotor() {
            _talon0.set(ControlMode.PercentOutput, 0);
            _talon1.set(ControlMode.PercentOutput, 0);
            _talon2.set(ControlMode.PercentOutput, 0);
        }

        @Override
        public double get() {
            return _talon0.getMotorOutputPercent();
        }
    };

    private MotorController _rightMotors = new MotorController() {
        @Override
        public void set(double speed) {
            _talon3.set(ControlMode.PercentOutput, speed);
            _talon4.set(ControlMode.PercentOutput, speed);
            _talon5.set(ControlMode.PercentOutput, speed);
        }

        @Override
        public void setInverted(boolean isInverted) {
            _talon3.setInverted(true);
            _talon4.setInverted(true);
            _talon5.setInverted(true);
        }

        @Override
        public boolean getInverted() {
            return _talon3.getInverted();
        }

        @Override
        public void disable() {
        }

        @Override
        public void stopMotor() {
            _talon3.set(ControlMode.PercentOutput, 0);
            _talon4.set(ControlMode.PercentOutput, 0);
            _talon5.set(ControlMode.PercentOutput, 0);
        }

        @Override
        public double get() {
            return _talon0.getMotorOutputPercent();
        }
    };

    private MotorControllerGroup m_LeftMotors = new MotorControllerGroup(_leftMotors);
    private MotorControllerGroup m_RightMotors = new MotorControllerGroup(_rightMotors);

    private DifferentialDrive m_Drive = new DifferentialDrive(m_LeftMotors,m_RightMotors);

    private XboxController m_Controller = new XboxController(0);

    private boolean m_LimelightHasValidTarget = false;
    private double m_LimelightDriveCommand = 0.0;
    private double m_LimelightSteerCommand = 0.0;

    // Joystick (for teleop)
    Joystick _joystick = new Joystick(0);

    // Autonomous variables
    private Timer autoTimer = new Timer();
    private final double AUTO_TIME = 15.0; // Autonomous period in seconds
    private final double TARGET_DISTANCE = 5.0; // Target distance in feet
    private boolean reachedTarget = false;

    // Network Tables for Limelight
    NetworkTable limelightTable;

    @Override
    public void robotInit() {
        // Initialize Network Tables
        limelightTable = NetworkTableInstance.getDefault().getTable("limelight");
        

        // Robot initialization code here (if any)

        CameraServer.startAutomaticCapture();

        m_chooser.setDefaultOption("Default Auto", kDefaultAuto);
        m_chooser.addOption("My Auto", kCustomAuto);
        SmartDashboard.putData("Auto choices", m_chooser);
    }

    @Override
    public void robotPeriodic() {
    }

    @Override
    public void autonomousInit() {
        // Reset and start the autonomous timer
        autoTimer.reset();
        autoTimer.start();
        reachedTarget = false;

        // Initial motor configuration for autonomous
        _talon0.setInverted(false);
        _talon2.setInverted(false);
        _talon4.setInverted(false);
        _talon1.setInverted(true);
        _talon3.setInverted(true);
        _talon5.setInverted(true);

        m_autoSelected = m_chooser.getSelected();
    }

    @Override
    public void autonomousPeriodic() {
        // Check if the autonomous period is still active and the target hasn't been reached
        // if (autoTimer.get() < AUTO_TIME && !reachedTarget) {
            // Get Limelight data
            NetworkTableEntry tv = limelightTable.getEntry("tv"); // Whether the limelight has any valid targets (0 or 1)
            NetworkTableEntry tx = limelightTable.getEntry("tx"); // Horizontal Offset From Crosshair To Target (-27 degrees to 27 degrees)
            NetworkTableEntry ty = limelightTable.getEntry("ty"); // Vertical Offset From Crosshair To Target (-20.5 degrees to 20.5 degrees)
            double targetVisible = tv.getDouble(0);
            double targetX = tx.getDouble(0);
            double targetY = ty.getDouble(0);
            SmartDashboard.putNumber("TX", targetX);
            SmartDashboard.putNumber("TV", targetVisible);
            // if (targetVisible) {
             
                double steer = targetX * 0.3;
                double drive = -0.02 * targetY;
                drive = Math.max(drive, 0.4);

                double leftCommand = drive + steer;
                double rightCommand = drive - steer;

                _talon0.set(ControlMode.PercentOutput, leftCommand);
                _talon1.set(ControlMode.PercentOutput, leftCommand);
                _talon2.set(ControlMode.PercentOutput, leftCommand);
                _talon3.set(ControlMode.PercentOutput, rightCommand);
                _talon4.set(ControlMode.PercentOutput, rightCommand);
                _talon5.set(ControlMode.PercentOutput, rightCommand);
            // } else {
                // Stop or search for target
                // stopMotors();
            // }

        // } else {
            // Stop the robot
            // stopMotors();
        // }
    }

    void stopMotors() {
      _talon0.set(ControlMode.PercentOutput, 0);
      _talon1.set(ControlMode.PercentOutput, 0);
      _talon2.set(ControlMode.PercentOutput, 0);
      _talon3.set(ControlMode.PercentOutput, 0);
      _talon4.set(ControlMode.PercentOutput, 0);
      _talon5.set(ControlMode.PercentOutput, 0);
    }

    @Override
    public void teleopPeriodic() {
        Update_Limelight_Tracking();

        double steer = m_Controller.getRawAxis(4) * 0.3;
        double drive = m_Controller.getRawAxis(1) * 0.3;
        boolean auto = m_Controller.getAButton();

        steer *= 0.70;
        drive *= 0.70;

        if (auto)
        {
            if (m_LimelightHasValidTarget)
            {
                    // m_Drive.arcadeDrive(m_LimelightDriveCommand,m_LimelightSteerCommand);

                    double left = m_LimelightDriveCommand + m_LimelightSteerCommand;
                    double right = m_LimelightDriveCommand - m_LimelightSteerCommand;

                    _leftMotors.set(left);
                    _rightMotors.set(right);
            }
            else
            {
                    // m_Drive.arcadeDrive(0.0,0.0);
                    _leftMotors.set(0);
                    _rightMotors.set(0);
            }
        }
        else
        {
            double left = drive + steer;
            double right = drive - steer;
            
            _leftMotors.set(left);
            _rightMotors.set(-right);
        }
    }

    public void Update_Limelight_Tracking()
    {
        // These numbers must be tuned for your Robot!  Be careful!
        final double STEER_K = 0.03;                    // how hard to turn toward the target
        final double DRIVE_K = 0.26;                    // how hard to drive fwd toward the target
        final double DESIRED_TARGET_AREA = 13.0;        // Area of the target when the robot reaches the wall
        final double MAX_DRIVE = 0.7;                   // Simple speed limit so we don't drive too fast

        double tv = NetworkTableInstance.getDefault().getTable("limelight").getEntry("tv").getDouble(0);
        double tx = NetworkTableInstance.getDefault().getTable("limelight").getEntry("tx").getDouble(0);
        double ty = NetworkTableInstance.getDefault().getTable("limelight").getEntry("ty").getDouble(0);
        double ta = NetworkTableInstance.getDefault().getTable("limelight").getEntry("ta").getDouble(0);

        if (tv < 1.0)
        {
          m_LimelightHasValidTarget = false;
          m_LimelightDriveCommand = 0.0;
          m_LimelightSteerCommand = 0.0;
          return;
        }

        m_LimelightHasValidTarget = true;

        double steer_cmd = tx * STEER_K;
        m_LimelightSteerCommand = steer_cmd;

        double drive_cmd = (DESIRED_TARGET_AREA - ta) * DRIVE_K;

        if (drive_cmd > MAX_DRIVE)
        {
          drive_cmd = MAX_DRIVE;
        }
        m_LimelightDriveCommand = drive_cmd;
  }
}

File: robot.py
#!/usr/bin/env python3
import wpilib
from wpilib.drive import DifferentialDrive
from wpilib import XboxController, Timer
from wpilib import SendableChooser, SmartDashboard
import networktables

class Robot(wpilib.TimedRobot):
    def robotInit(self):
        # Constants
        self.kDefaultAuto = "Default"
        self.kCustomAuto = "My Auto"

        # Motor controllers
        self.talon0 = wpilib.Talon(1)
        self.talon1 = wpilib.Talon(3)
        self.talon2 = wpilib.Talon(5)
        self.talon3 = wpilib.Talon(2)
        self.talon4 = wpilib.Talon(4)
        self.talon5 = wpilib.Talon(6)

        def set_left_motors(speed):
            self.talon0.set(speed)
            self.talon1.set(speed)
            self.talon2.set(speed)

        def set_right_motors(speed):
            self.talon3.set(speed)
            self.talon4.set(speed)
            self.talon5.set(speed)

        # Drivetrain
        self.drive = DifferentialDrive(set_left_motors, set_right_motors)

        # Controllers
        self.controller = XboxController(0)

        # Limelight variables
        self.limelight_has_valid_target = False
        self.limelight_drive_command = 0.0
        self.limelight_steer_command = 0.0

        # Autonomous variables
        self.auto_timer = Timer()
        self.auto_time = 15.0
        self.target_distance = 5.0
        self.reached_target = False

        # Network Tables for Limelight
        self.limelight_table = networktables.NetworkTablesInstance.getDefault().getTable("limelight")

        # Autonomous selection
        self.chooser = SendableChooser()
        self.chooser.setDefaultOption("Default Auto", self.kDefaultAuto)
        self.chooser.addOption("My Auto", self.kCustomAuto)
        SmartDashboard.putData("Auto choices", self.chooser)

    def autonomousInit(self):
        self.auto_timer.reset()
        self.auto_timer.start()
        self.reached_target = False

        self.talon0.setInverted(False)
        self.talon2.setInverted(False)
        self.talon4.setInverted(False)
        self.talon1.setInverted(True)
        self.talon3.setInverted(True)
        self.talon5.setInverted(True)

        self.auto_selected = self.chooser.getSelected()

    def teleopPeriodic(self):
        self.update_limelight_tracking()

        steer = self.controller.getRawAxis(4) * 0.3
        drive = self.controller.getRawAxis(1) * 0.3
        auto = self.controller.getAButton()

        if auto:
            if self.limelight_has_valid_target:
                left = self.limelight_drive_command + self.limelight_steer_command
                right = self.limelight_drive_command - self.limelight_steer_command

                self.talon0.set(left)
                self.talon1.set(left)
                self.talon2.set(left)
                self.talon3.set(-right)
                self.talon4.set(-right)
                self.talon5.set(-right)
            else:
                self.talon0.set(0)
                self.talon1.set(0)
                self.talon2.set(0)
                self.talon3.set(0)
                self.talon4.set(0)
                self.talon5.set(0)
        else:
            left = drive + steer
            right = drive - steer

            self.talon0.set(left)
            self.talon1.set(left)
            self.talon2.set(left)
            self.talon3.set(-right)
            self.talon4.set(-right)
            self.talon5.set(-right)

    def update_limelight_tracking(self):

        STEER_K = 0.03  # How hard to turn toward the target
        DRIVE_K = 0.26  # How hard to drive forward
        DESIRED_TARGET_AREA = 13.0  # Target area when the robot is at the wall
        MAX_DRIVE = 0.7  # Limit on drive speed

        # Get Limelight values (Check if the entry names match your setup)
        tv = self.limelight_table.getEntry("tv").getDouble(0.0)
        tx = self.limelight_table.getEntry("tx").getDouble(0.0)
        ty = self.limelight_table.getEntry("ty").getDouble(0.0)
        ta = self.limelight_table.getEntry("ta").getDouble(0.0)

        # Check for valid target
        if tv < 1.0:
            self.limelight_has_valid_target = False
            self.limelight_drive_command = 0.0
            self.limelight_steer_command = 0.0
            return  # Stop execution if no target is found

        # Target is found!
        self.limelight_has_valid_target = True

        # Calculate steering command
        steer_cmd = tx * STEER_K
        self.limelight_steer_command = steer_cmd

        # Calculate drive command
        drive_cmd = (DESIRED_TARGET_AREA - ta) * DRIVE_K
        if drive_cmd > MAX_DRIVE:
            drive_cmd = MAX_DRIVE
        self.limelight_drive_command = drive_cmd

if __name__ == "__main__":
    wpilib.run(Robot)
